src/idl/tests/annotation.c

```diff
@@ -942,6 +942,39 @@ CU_Test(idl_annotation, parameter_scope)
 #undef A
 #undef TA
 
+CU_Test(idl_annotation, identifier_clash)
+{
+  static const struct {
+    const char *str;
+    idl_retcode_t ret;
+  } tests[] = {
+    { "struct key { @key long f1; };", IDL_RETCODE_SEMANTIC_ERROR },
```

Comment:
Is this really correct?

In XTypes 1.3 it says:

> 7.3.1.2.4 Defining Annotations
> Annotation types shall be represented as described in this clause. An annotation type is defined using the new token @annotation, as in the following example:
> ```
> @annotation MyAnnotation { // ...
> };
> ```
> Annotation identifiers are orthogonal to any other kind of type and therefore do not conflict with other types that may use the same identifier name even when defined in the same module. This is because the application of an annotation prefixes the annotation identifier with the “@” character, see Sub Clause 7.3.1.2.5.

That suggests that defining a `struct key` is perfectly acceptable.

Reply:
You're right, I missed that section. In the updated commit I've rewritten the fix, which I think now correctly handles the cases with an annotation that has the same name as a type in the same scope. What still does not work is using annotations with a fully scoped name, e.g. `@module1::my_annotation` (and also: `@module1::key`, which I think is allowed). I'll create an issue for this.